docs/raster/S3Streaming.md

```diff
@@ -0,0 +1,118 @@
+## S3 GeoTiff Streaming
+
+- [Introduction](#introduction)
+- [How to Use thise Feature](#how-to-use-this-feature)
+  - [Creating BytesStreamer](#creating-streambytes)
+  - [Creating a ByteReader](#creating-a-bytereader)
+  - [Streaming a GeoTiff From S3](#streaming-a-geotiff-from-s3)
+    - [Streaming the Entire GeoTiff](#streaming-the-entire-geotiff)
+    - [Streaming a Section of a GeoTiff](#streaming-a-section-of-a-geotiff)
+- [Conclusion](#conclusion)
+
+- - -
+### Introduction
+
+A common problem that faces those who perform geo-spatial analysis using Amazon's S3 is dealing with importing and working with large GeoTiffs on the S3 server. A solution to this issue has been in development for the past few weeks that will allow for stream reading of GeoTiffs directly off S3 regardless of the file's size. This short guide goes over how to use this feature as of 10/12/2016.
+
+**Note** This feature is still be worked on and is subject to change at any time. I will update this document when any changes have been made to this feature as soon as I can. To check the latest version, see the link to the pull request at the bottom of this document.
+- - -
+### How to Use this Feature
+
+In order for streaming from S3 to work, two new types have been created. These are: `BytesStreamer` and `ByteReader`, respectively. the first deals with getting bytes in the form of `Map[Long, Array[Byte]` called a, "chunk". The `Long` of this chunk is the position in the file where the `Array[Byte]` starts. For example, if a new chunk needs to be created at the 100th byte of the file then the chunk that is created will have a value of 100. These chunks can vary in size and a new one is produced each time a call is made to another part of the file which is not contained within the current chunk.
+
+`ByteReader` is what reads the bytes from the chunks. As of right now, only `java.nio.ByteBuffer`s and the `StreamByteReader` type can access the chunk.
+
+#### Creating `BytesStreamer`
+There is only one way to construct an instance of `BytesStreamer` as of this writing, and that is through the `S3ByteStreamer` class. Found in `geotrellis.spark.io.s3.util` package, this class will be able to get chunks from a desired GeoTifff on S3. Below is a basic guide on creating this class:
+
+```scala
+ import geotrellis.spark.io.s3._
+ import geotrellis.spark.io.s3.util.S3BytesStreamer
+ 
+ import com.amazonaws.services.s3.model._
+ 
+ val client: S3Client = S3Client.default
+ val bucket: String = "my-bucket"
+ val key: String = "path/to/my/geotiff.tif"
+ 
+ // how many bytes that should be streamed at a time
+ val chunkSize: Int = 256000
+ 
+ val s3bytes = S3BytesStreamer(bucket, key, client, chunkSize)
```

Comment:
`S3BytesStreamer` not works with `S3Client` type, but works with `AmazonS3Client`, i think it is possible to use more generic `S3Client` type in `S3BytesStreamer` constructor.